torch/lib/c10d/Utils.hpp

```diff
@@ -3,18 +3,124 @@
 #include <sys/socket.h>
 #include <sys/types.h>
 
-#include <cstdlib>
+#include <chrono>
 #include <cstdint>
+#include <cstdlib>
 #include <functional>
 #include <limits>
 #include <string>
 #include <system_error>
 #include <tuple>
 #include <vector>
-#include <chrono>
+
+#include <ATen/ATen.h>
+
+#include "Types.hpp"
+
+inline void hash_combine(std::size_t& seed) {}
+
+template <typename T, typename... Rest>
+inline void hash_combine(std::size_t& seed, const T& v, Rest... rest) {
+  std::hash<T> hasher;
+  seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+  hash_combine(seed, rest...);
+}
+
+#define MAKE_HASHABLE(type, ...)                  \
+  namespace std {                                 \
+  template <>                                     \
+  struct hash<type> {                             \
+    std::size_t operator()(const type& t) const { \
+      std::size_t ret = 0;                        \
+      hash_combine(ret, __VA_ARGS__);             \
+      return ret;                                 \
+    }                                             \
+  };                                              \
+  }
+
+// Make std::vector<T> hashable
+namespace std {
+template <typename T>
+struct hash<vector<T>> {
+  std::size_t operator()(const vector<T>& v) const {
+    std::size_t ret = v.size();
+    for (const auto& e : v) {
+      hash_combine(ret, e);
+    }
+    return ret;
+  }
+};
+} // namespace std
+
+// Make at::ArrayRef<T> hashable
+namespace std {
+template <typename T>
+struct hash<at::ArrayRef<T>> {
+  std::size_t operator()(const at::ArrayRef<T>& v) const {
+    std::size_t ret = v.size();
+    for (const auto& e : v) {
+      hash_combine(ret, e);
+    }
+    return ret;
+  }
+};
+} // namespace std
+
+MAKE_HASHABLE(::c10d::CollectiveType, static_cast<std::uint8_t>(t));
+MAKE_HASHABLE(::c10d::DeviceType, static_cast<std::uint8_t>(t));
+MAKE_HASHABLE(::c10d::ReduceOp, static_cast<std::uint8_t>(t));
 
 namespace c10d {
 
+inline void assertSameSizeAndType(const std::vector<at::Tensor>& tensors) {
+  // Ensure we have at least one tensor
+  if (tensors.size() == 0) {
+    throw std::invalid_argument("argument is empty");
+  }
+
+  // Ensure all tensors have identical type and shape
+  auto& type = tensors[0].type();
+  auto sizes = tensors[0].sizes();
+  for (auto i = 1; i < tensors.size(); i++) {
+    if (tensors[i].type() != type) {
+      throw std::invalid_argument("argument contains mixed types");
+    }
+    if (!tensors[i].sizes().equals(sizes)) {
+      throw std::invalid_argument("argument contains mixed sizes");
+    }
+  }
+}
```

Comment:
We should make this function accept a caller name. Getting an error message like "argument contains mixed sizes" is very unhelpful. Mentioning what the expected/received size is would make it even better.

Reply:
Added argument values to compare to error message.

For types I'm using at::Type#toString().

For sizes I have an inline toString function that outputs "(1, 2, 3, 4)".